### ISSUE TITLE
Automatically wrap dicts in config with `Field(Dict)`

### ISSUE THREAD

**Issue by helloworld** [dagster-io/dagster#1908]
If I want to have nested dictionaries in my config, I have to define my config like:

```
config_field=Field(Dict({'hyperparameters': Field(Dict({'dropout': Field(Float), 'learning_rate': Field(Float)}))}))
```

We already have the shorthand `config` argument that takes in a literal `dict` and wraps it in `Field(Dict)`:

```
config={'hyperparameters': Field(Dict({'dropout': Field(Float), 'learning_rate': Field(Float)}))}
```

It would be nice if all `dict`s were recursively wrapped in `Field(Dict)`. The ideal syntax would be:

```
config={'hyperparameters': {'dropout': Float, 'learning_rate': Float}}
```

**Comment by schrockn** [comment]
Going down this road of automatic type coercion everywhere is one of those things that seems to simplify things in the short turn but can inject a lot of complexity and confusion. See how graphene coerces between Fields and Types all over the place. Seems good in the short term but in the end it is quite a mess.

Any explicit wrapper function could be reasonable. That accepts the raw dict and produces the full dictionary with no default and all required fields.

If you open this API up, then the next thing is that someone says "hey but I want learning_rate to be optional".

Then you allow:

```{'hyperparameters': {'dropout': Float, 'learning_rate': Field(Float, is_optional=True)}}```

So you are doing a recursive descent now and coercing and not coercing things. It ends up being really confusing.

**Comment by mgasner** [comment]
I already argued for `config` and continue to think that something like this is what most of our users will want to operate in

**Comment by mgasner** [comment]
#1777

**Comment by alangenfeld** [comment]
> Then you allow `{'hyperparameters': {'dropout': Float, 'learning_rate': Field(Float, is_optional=True)}}`
> ...
> ends up being really confusing

Hmm this actually feels reasonable to me. Only needing to use `Field` when I need to express information that is not just the type seems like a decent gradual disclosure of complexity. 

I don't think it's clear cut whether the implementation would be that complex or that this is a problematic degree of coercion / inference.

I think an RFC diff would be best for further discussing the trade offs here.

**Comment by asingh16** [comment]
More evidence for why I think Fields should only be used for the expression of information (or default behavior) is in training solids. Here is a solid where I am configuring the architecture for a neural network, this feels unnecessarily verbose. This makes me feel like I am writing LISP:
```
@solid(
    config={
        'lstm_layer_config': Field(Dict({
            'activation': Field(String, is_optional=True, default_value='relu')
        })),
        'model_trainig_config': Field(Dict({
            'optimizer': Field(String, description='Type of optimizer to use', is_optional=True, default_value='adam'),
            'loss': Field(String, is_optional=True, default_value='mse'),
            'num_epochs': Field(Int, description='Number of epochs to optimize over',),
        })),
    }
)
```

**Comment by schrockn** [comment]
What would be your ideal way to express the same amount of information you are expressing here?

**Comment by schrockn** [comment]
```
def field_dict(fields):
    return Field(Dict(fields)) 

@solid(
    config={
        'lstm_layer_config': field_dict({
            'activation': Field(String, is_optional=True, default_value='relu')
        }),
        'model_trainig_config': field_dict({
            'optimizer': Field(String, description='Type of optimizer to use', is_optional=True, default_value='adam'),
            'loss': Field(String, is_optional=True, default_value='mse'),
            'num_epochs': Field(Int, description='Number of epochs to optimize over',),
        }),
    }
)
```

like that?

**Comment by asingh16** [comment]
So you are definitely right that we ought to not try to build something that supports everything but rather force an explicit tradeoff (For example the "|" being used to overload solids in the case of super simple pipelines). If you have a config that is dead simple and JUST need type validation at the leaves of the config, then I want to express my config as the following:

```
config={
    'lstm_layer_config': {
            'activation': Field(String, is_optional=True, default_value='relu')
        },
    'model_trainig_config': {
            'optimizer': Field(String, description='Type of optimizer to use', is_optional=True, default_value='adam'),
            'loss': Field(String, is_optional=True, default_value='mse'),
            'num_epochs': Field(Int, description='Number of epochs to optimize over',),
      },
}
```

**Comment by asingh16** [comment]
If the user wants to use complex selector logic and do defaulting at non leaf config keys, then they have to go the verbose route, but I assume that a majority of users will not need anything more than the above.

**Comment by schrockn** [comment]
Been convinced that this is worth considering. will prototype

**Comment by schrockn** [comment]
https://dagster.phacility.com/D1614

**Comment by schrockn** [comment]
done!
